### ISSUE TITLE
ERROR when running keylime_agent

### ISSUE THREAD

**Issue by hustliyilin** [keylime/rust-keylime#586]
When I run the rust-keylime, I get the following logs

```shell
# keylime_agent
WARNING:esys:src/tss2-esys/api/Esys_NV_ReadPublic.c:309:Esys_NV_ReadPublic_Finish() Received TPM Error
ERROR:esys:src/tss2-esys/esys_tr.c:243:Esys_TR_FromTPMPublic_Finish() Error NV_ReadPublic ErrorCode (0x0000018b)
ERROR:esys:src/tss2-esys/esys_tr.c:398:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x0000018b)
 ERROR tss_esapi::context::general_esys_tr > Error when getting ESYS handle from TPM handle: the handle is not correct for the use (associated with handle number 1)
```

![image](https://github.com/keylime/rust-keylime/assets/6824093/da7fdf02-b7b6-4f13-a87d-27826c5d9ef1)

On the keylime_registrar side, It can still get a post request about activating agent like this:

![image](https://github.com/keylime/rust-keylime/assets/6824093/37185203-e78d-43cf-8450-8c4b7a23386f)

How can I fix the keylime_agent errors above? Could you help me? Thanks a lot.

**Comment by ansasaki** [comment]
Which version of `tpm2-tss` llibrary are you using? Also, I see you are using an old version  of `tpm2-tools`. I remember having similar issues, but I couldn't find exactly what was the issue. Please try with newer versions of `tpm2-tools` and `tpm2-tss`.

**Comment by hustliyilin** [comment]
Thanks a lot for your quick reply.
1. I deployed the agent on a single machine. Currently, the agent already uses the latest `tpm2-tss` and `tpm2-tools`.

![image](https://github.com/keylime/rust-keylime/assets/6824093/95ae4a9e-0b65-480e-bdfe-36e152540ae9)

2. The older tpm2-tools version(`version 4.1.1`) of my registrar is on another machine. And I just tried it, and the log is similar even if the registrar  is not started.
3. I just found that the error is reported when executing [retrieve_ek_pubcert](https://github.com/keylime/rust-keylime/blob/master/keylime/src/tpm.rs#L175) to obtain the ek certificate. After the error is reported, the debug log will print `No EK certificate found in TPM NVRAM`, which should be a TPM-related error.

**Comment by ansasaki** [comment]
Could you please try running the keylime agent with debug logs to that we can see all messages? Try:
```
$ RUST_LOG=debug keylime_agent
```

**Comment by hustliyilin** [comment]
OK, This is my debug logs with the `RUST_LOG=debug keylime_agent` command:

```shell
# RUST_LOG=debug keylime_agent
 DEBUG keylime_agent::secure_mount > Secure store location /var/lib/keylime/secure already mounted on tmpfs
 INFO  keylime_agent::permissions  > Changed file /var/lib/keylime/secure owner to keylime:tss.
 INFO  keylime_agent::permissions  > Dropped privileges to run as keylime:tss
 INFO  keylime_agent               > Running the service as keylime:tss...
 INFO  keylime_agent               > Starting server with API version v2.0...
 WARN  keylime_agent               > INSECURE: Keylime is using a software TPM emulator rather than a real hardware TPM.
 WARN  keylime_agent               > INSECURE: The security of Keylime is NOT linked to a hardware root of trust.
 WARN  keylime_agent               > INSECURE: Only use Keylime in this mode for testing or debugging purposes.
WARNING:esys:src/tss2-esys/api/Esys_NV_ReadPublic.c:309:Esys_NV_ReadPublic_Finish() Received TPM Error
ERROR:esys:src/tss2-esys/esys_tr.c:243:Esys_TR_FromTPMPublic_Finish() Error NV_ReadPublic ErrorCode (0x0000018b)
ERROR:esys:src/tss2-esys/esys_tr.c:398:Esys_TR_FromTPMPublic() Error TR FromTPMPublic ErrorCode (0x0000018b)
 ERROR tss_esapi::context::general_esys_tr > Error when getting ESYS handle from TPM handle: the handle is not correct for the use (associated with handle number 1)
 WARN  keylime::tpm                        > No EK certificate found in TPM NVRAM
 INFO  keylime_agent                       > Loaded old AK key from /var/lib/keylime/agent_data.json
 INFO  keylime_agent                       > Agent UUID: d432fbb3-d2f1-4a97-9ef7-75bd81c00000
 DEBUG keylime_agent                       > Loading existing key pair from /var/lib/keylime/server-private.pem
 DEBUG keylime_agent                       > Loading existing mTLS certificate from /var/lib/keylime/server-cert.crt
 INFO  keylime_agent::registrar_agent      > Requesting agent registration from http://8.217.96.172:8890/v2.0/agents/d432fbb3-d2f1-4a97-9ef7-75bd81c00000 for d432fbb3-d2f1-4a97-9ef7-75bd81c00000
 DEBUG reqwest::connect                    > starting new connection: http://8.217.96.172:8890/
 DEBUG hyper::client::connect::http        > connecting to 8.217.96.172:8890
 DEBUG hyper::client::connect::http        > connected to 8.217.96.172:8890
 DEBUG hyper::proto::h1::io                > flushed 2110 bytes
 DEBUG hyper::proto::h1::io                > parsed 3 headers
 DEBUG hyper::proto::h1::conn              > incoming body is close-delimited
 DEBUG hyper::proto::h1::conn              > incoming body completed
 INFO  keylime_agent                       > SUCCESS: Agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 registered
 INFO  keylime_agent::registrar_agent      > Requesting agent activation from http://8.217.96.172:8890/v2.0/agents/d432fbb3-d2f1-4a97-9ef7-75bd81c00000 for d432fbb3-d2f1-4a97-9ef7-75bd81c00000
 DEBUG reqwest::connect                    > starting new connection: http://8.217.96.172:8890/
 DEBUG hyper::client::connect::http        > connecting to 8.217.96.172:8890
 DEBUG hyper::client::connect::http        > connected to 8.217.96.172:8890
 DEBUG hyper::proto::h1::io                > flushed 268 bytes
 DEBUG hyper::proto::h1::io                > parsed 3 headers
 DEBUG hyper::proto::h1::conn              > incoming body is close-delimited
 DEBUG hyper::proto::h1::conn              > incoming body completed
 INFO  keylime_agent                       > SUCCESS: Agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 activated
 INFO  actix_server::builder               > Starting 2 workers
 INFO  keylime_agent                       > Listening on https://0.0.0.0:9002
 DEBUG keylime_agent::revocation           > Starting revocation worker
 INFO  actix_server::server                > Actix runtime found; starting in Actix runtime
 DEBUG keylime_agent::payloads             > Starting payloads worker
 DEBUG keylime_agent::keys_handler         > Starting keys worker
```

![image](https://github.com/keylime/rust-keylime/assets/6824093/06ef3002-eb3a-4dc4-9d32-9fc92043651a)

Could you help me? Thanks.

**Comment by ansasaki** [comment]
I cannot reproduce your issue. Apparently, the TPM you are using, for some reason, does not store the EK certificate in the default address (NV Index `0x01c00002`).  That address is specified in the [TCG EK Credential Profile for TPM Family 2.0](https://trustedcomputinggroup.org/wp-content/uploads/TCG-EK-Credential-Profile-V-2.5-R2_published.pdf), section 2.2.1.4

Could you try to run the following commands in that machine:
```
$ tpm2_getekcertificate > ek.crt
$ openssl x509 -in ek.crt -inform DER -text
```

That should print the EK certificate of your TPM. I see you are using a software TPM, make sure to set the correct TCTI for the `tpm2_getekcertificate` command by setting the `TPM2TOOLS_TCTI` env var.

If you know that your TPM uses a non-default NV index for the EK certificate, use the `ek_handle` configuration option to set the correct EK handle. You can also do that by setting the `KEYLIME_AGENT_EK_HANDLE` env var.

**Comment by hustliyilin** [comment]
Thanks for your reply,  When I executed the command you provided, I found that the EK public key also needs to be specified. 

![8d20f52b08dcbcd22321d8ed860ada6c](https://github.com/keylime/rust-keylime/assets/6824093/b265fb6a-0a90-41e4-8d9d-4f4e4e514a11)

But I don’t know the location of the EK public key at the moment. This should be My TPM's problem rather than keylime's problem.  I will close this issue first.  Thanks again for your help.
